Nama : Aditya Atadewa  
Kelas : TI 3G  
NIM : 2341720174  
Absen : 01  

# Tugas Praktikum

Menggunakan JST untuk klasifikasi angka tulisan tangan (MNIST).

## Langkah:

1. Load dataset MNIST dari Keras.
2. Bangun model dengan 2 hidden layer.
3. Latih model dan evaluasi akurasi.

In [1]:
import tensorflow as tf
from tensorflow.keras.datasets import mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Flatten
from tensorflow.keras.utils import to_categorical
import time

# Load MNIST
(X_train, y_train), (X_test, y_test) = mnist.load_data()

# Normalisasi
X_train = X_train / 255.0
X_test = X_test / 255.0

# One-hot label
y_train = to_categorical(y_train, 10)
y_test = to_categorical(y_test, 10)

# Model baseline (2 hidden layer)
model = Sequential([
    Flatten(input_shape=(28,28)),
    Dense(128, activation='relu'),
    Dense(64, activation='relu'),
    Dense(10, activation='softmax')
])

# Compile
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

# Train + ukur waktu
start = time.time()
history = model.fit(X_train, y_train, epochs=5, batch_size=32, verbose=1)
end = time.time()

# Evaluasi
loss, acc = model.evaluate(X_test, y_test)
print(f"Akurasi Baseline: {acc:.4f}")
print(f"Waktu Training: {end - start:.2f} detik")

11490434/11490434 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step


/usr/local/lib/python3.12/dist-packages/keras/src/layers/reshaping/flatten.py:37: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 1/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step - accuracy: 0.8830 - loss: 0.4117
Epoch 2/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 6s 3ms/step - accuracy: 0.9693 - loss: 0.1018
Epoch 3/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 10s 3ms/step - accuracy: 0.9786 - loss: 0.0677
Epoch 4/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 6s 3ms/step - accuracy: 0.9846 - loss: 0.0501
Epoch 5/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.9867 - loss: 0.0402
313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9704 - loss: 0.0944
Akurasi Baseline: 0.9740
Waktu Training: 35.75 detik


**Struktur baseline:**
- Dense 128 (ReLU)
- Dense 64 (ReLU)
- Output 10 (Softmax)

**Hasil baseline:**
- **Akurasi:** `0.9740`
- **Waktu training:** `35.75 detik`

### Soal 1 – Ubah jumlah neuron (256 dan 128)

In [2]:
model = Sequential([
    Flatten(input_shape=(28,28)),
    Dense(256, activation='relu'),
    Dense(128, activation='relu'),
    Dense(10, activation='softmax')
])

model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

start = time.time()
history = model.fit(X_train, y_train, epochs=5, batch_size=32, verbose=1)
end = time.time()

loss, acc = model.evaluate(X_test, y_test)
print(f"Akurasi Neuron 256-128: {acc:.4f}")
print(f"Waktu Training: {end - start:.2f} detik")


Epoch 1/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 10s 5ms/step - accuracy: 0.8926 - loss: 0.3595
Epoch 2/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 10s 5ms/step - accuracy: 0.9726 - loss: 0.0877
Epoch 3/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step - accuracy: 0.9805 - loss: 0.0592
Epoch 4/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step - accuracy: 0.9869 - loss: 0.0416
Epoch 5/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 8s 5ms/step - accuracy: 0.9889 - loss: 0.0322
313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9709 - loss: 0.0993
Akurasi Neuron 256-128: 0.9763
Waktu Training: 44.62 detik


### **Konfigurasi:**
- Dense 256 (ReLU)
- Dense 128 (ReLU)
- Output 10

### **Hasil:**
- **Akurasi:** `0.9763`
- **Waktu training:** `44.62 detik`

### **Kesimpulan:**

- Menambah jumlah neuron menjadi 256–128 sedikit meningkatkan akurasi menjadi **97.63%**, namun waktu training meningkat cukup signifikan menjadi **44.62 detik**.  
- Model menjadi lebih kompleks sehingga membutuhkan waktu lebih lama.

## Soal 2 - Tambahkan 1 hidden layer lagi

In [3]:
model = Sequential([
    Flatten(input_shape=(28,28)),
    Dense(128, activation='relu'),
    Dense(64, activation='relu'),
    Dense(32, activation='relu'),    # hidden layer tambahan
    Dense(10, activation='softmax')
])

model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

start = time.time()
history = model.fit(X_train, y_train, epochs=5, batch_size=32, verbose=1)
end = time.time()

loss, acc = model.evaluate(X_test, y_test)
print(f"Akurasi 3 Hidden Layer: {acc:.4f}")
print(f"Waktu Training: {end - start:.2f} detik")


Epoch 1/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 8s 3ms/step - accuracy: 0.8578 - loss: 0.4701
Epoch 2/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 6s 3ms/step - accuracy: 0.9663 - loss: 0.1110
Epoch 3/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 6s 3ms/step - accuracy: 0.9765 - loss: 0.0755
Epoch 4/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 6s 3ms/step - accuracy: 0.9819 - loss: 0.0575
Epoch 5/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.9848 - loss: 0.0475
313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.9718 - loss: 0.0974
Akurasi 3 Hidden Layer: 0.9762
Waktu Training: 31.21 detik


### **Konfigurasi:**
- Dense 128 (ReLU)
- Dense 64 (ReLU)
- Dense 32 (ReLU) ← tambahan
- Output 10

### **Hasil:**
- **Akurasi:** `0.9762`
- **Waktu training:** `31.21 detik`

### **Kesimpulan:**

- Penambahan satu hidden layer tambahan memberikan sedikit peningkatan akurasi (menjadi **97.62%**) dengan waktu training yang **lebih cepat** dibanding baseline.  
- Layer tambahan membantu representasi fitur tanpa menambah waktu komputasi terlalu besar.

## Soal 3 – Aktivasi Sigmoid vs ReLU

### Model dengan Sigmoid

In [4]:
model = Sequential([
    Flatten(input_shape=(28,28)),
    Dense(128, activation='sigmoid'),
    Dense(64, activation='sigmoid'),
    Dense(10, activation='softmax')
])

model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

start = time.time()
history = model.fit(X_train, y_train, epochs=5, batch_size=32, verbose=1)
end = time.time()

loss, acc = model.evaluate(X_test, y_test)
print(f"Akurasi Sigmoid: {acc:.4f}")
print(f"Waktu Training: {end - start:.2f} detik")


Epoch 1/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 7s 3ms/step - accuracy: 0.7800 - loss: 0.8684
Epoch 2/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 6s 3ms/step - accuracy: 0.9439 - loss: 0.1945
Epoch 3/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 6s 3ms/step - accuracy: 0.9627 - loss: 0.1297
Epoch 4/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.9719 - loss: 0.0944
Epoch 5/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 6s 3ms/step - accuracy: 0.9797 - loss: 0.0719
313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9644 - loss: 0.1091
Akurasi Sigmoid: 0.9693
Waktu Training: 30.43 detik


### Model dengan ReLU

In [5]:
model = Sequential([
    Flatten(input_shape=(28,28)),
    Dense(128, activation='relu'),
    Dense(64, activation='relu'),
    Dense(10, activation='softmax')
])

model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

start = time.time()
history = model.fit(X_train, y_train, epochs=5, batch_size=32, verbose=1)
end = time.time()

loss, acc = model.evaluate(X_test, y_test)
print(f"Akurasi ReLU: {acc:.4f}")
print(f"Waktu Training: {end - start:.2f} detik")


Epoch 1/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 7s 3ms/step - accuracy: 0.8803 - loss: 0.4189
Epoch 2/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 6s 3ms/step - accuracy: 0.9681 - loss: 0.1058
Epoch 3/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.9778 - loss: 0.0720
Epoch 4/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 6s 3ms/step - accuracy: 0.9844 - loss: 0.0493
Epoch 5/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.9874 - loss: 0.0396
313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9726 - loss: 0.0932
Akurasi ReLU: 0.9772
Waktu Training: 29.67 detik


#### **Model A – Sigmoid**
- Akurasi: `0.9693`
- Waktu training: `30.43 detik`

#### **Model B – ReLU**
- Akurasi: `0.9772`
- Waktu training: `29.67 detik`

#### **Kesimpulan**

- Fungsi aktivasi **ReLU** secara konsisten memberikan akurasi lebih tinggi (**97.72%**) dan training lebih cepat dibanding Sigmoid (**96.93%**).  
- ReLU lebih cocok untuk dataset besar seperti MNIST karena menghindari *vanishing gradient*.


## Perbandingan Semua Model


| Model / Tugas | Hidden Layer | Aktivasi | Akurasi | Waktu Training |
|---------------|--------------|----------|----------|----------------|
| Baseline | 128 → 64 | ReLU | **0.9740** | 35.75s |
| Tugas 1 | 256 → 128 | ReLU | **0.9763** | 44.62s |
| Tugas 2 | 128 → 64 → 32 | ReLU | **0.9762** | 31.21s |
| Tugas 3A | 128 → 64 | Sigmoid | **0.9693** | 30.43s |
| Tugas 3B | 128 → 64 | ReLU | **0.9772** | 29.67s |

Dari tabel di atas terlihat bahwa setiap perubahan arsitektur atau fungsi aktivasi memberikan dampak berbeda terhadap akurasi dan waktu pelatihan:

1. **Baseline (128 → 64, ReLU)**  
   Menjadi acuan. Akurasi sangat baik (**97.40%**) dengan waktu pelatihan sedang (35.75 detik).

2. **Tugas 1 (256 → 128, ReLU)**  
   Jumlah neuron lebih banyak membuat model lebih kompleks.  
   Hasilnya akurasi sedikit meningkat (**97.63%**), tetapi waktu pelatihan menjadi paling lama (**44.62 detik**) karena lebih banyak parameter.

3. **Tugas 2 (128 → 64 → 32, ReLU)**  
   Menambahkan satu hidden layer memperkaya representasi fitur.  
   Akurasi hampir terbaik (**97.62%**) dengan waktu pelatihan yang lebih cepat (**31.21 detik**).  
   Hal tersebut menunjukkan bahwa lebih banyak layer tidak selalu memperlambat model jika jumlah neuron per layer optimal.

4. **Tugas 3A (128 → 64, Sigmoid)**  
   Aktivasi Sigmoid menghasilkan akurasi terendah (**96.93%**) dan cenderung belajar lebih lambat.  
   Hal tersebut dapat terjadi karena Sigmoid rentan mengalami *vanishing gradient* terutama pada dataset besar seperti MNIST.

5. **Tugas 3B (128 → 64, ReLU)**  
   ReLU kembali menjadi yang terbaik dengan akurasi tertinggi (**97.72%**) dan waktu pelatihan tercepat (**29.67 detik**).  
   Hal tersebut menegaskan bahwa ReLU memang pilihan optimal untuk model deep learning modern.

Secara keseluruhan, **ReLU consistently outperform Sigmoid**, dan konfigurasi paling efisien adalah model dengan **2 hidden layer ReLU berukuran sedang**.